akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -17,159 +18,94 @@ import akka.util.OptionVal
+  private val _trueStatePredicate: FI.TypedPredicate[Any] = new FI.TypedPredicate[Any] {
+    override def defined(t: Any): Boolean = true
+  }
 
-  private var cases: List[CommandHandlerCase] = Nil
+  private def trueStatePredicate[S]: FI.TypedPredicate[S] = _trueStatePredicate.asInstanceOf[FI.TypedPredicate[S]]
 
-  private def addCase(predicate: Command ⇒ Boolean, handler: CommandToEffect[Command, Command, Event, State]): Unit = {
-    cases = CommandHandlerCase(predicate, handler) :: cases
-  }
   /**
-   * Match any command which the given `predicate` returns true for
+   * @param stateClass The handlers defined by this builder are used when the state is an instance of the `stateClass`
+   * @return A new, mutable, command handler builder
    */
-  def matchCommand(predicate: FI.TypedPredicate[Command], commandToEffect: CommandToEffect[Command, Command, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
-    addCase(cmd ⇒ predicate.defined(cmd), commandToEffect)
-    this
-  }
-
-  def matchCommand[C <: Command](commandClass: Class[C], commandToEffect: CommandToEffect[Command, C, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
-    addCase(cmd ⇒ commandClass.isAssignableFrom(cmd.getClass), commandToEffect.asInstanceOf[CommandToEffect[Command, Command, Event, State]])
-    this
-  }
+  def builder[Command, Event, S <: State, State](stateClass: Class[S]): CommandHandlerBuilder[Command, Event, S, State] =
+    new CommandHandlerBuilder(stateClass, statePredicate = trueStatePredicate)
 
   /**
-   * Builds a Command Handler and resets this builder
+   * @param statePredicate The handlers defined by this builder are used when the `statePredicate` is `true`,
+   *                       useful for example when state type is an Optional
+   * @return A new, mutable, command handler builder
    */
-  def build(): CommandHandler[Command, Event, State] = {
-    val builtCases = cases.reverse.toArray
-    cases = Nil
-    new CommandHandler[Command, Event, State] {
-      override def apply(ctx: ActorContext[Command], state: State, command: Command) = {
-        var idx = 0
-        var effect: OptionVal[Effect[Event, State]] = OptionVal.None
-
-        while (idx < builtCases.length && effect.isEmpty) {
-          val curr = builtCases(idx)
-          if (curr.predicate(command)) {
-            effect = OptionVal.Some(curr.handler.apply(ctx, state, command))
-          }
-          idx += 1
-        }
-
-        effect match {
-          case OptionVal.None    ⇒ throw new MatchError(s"No match found for command of type [${command.getClass.getName}]")
-          case OptionVal.Some(e) ⇒ e.asInstanceOf[EffectImpl[Event, State]]
-        }
-      }
-    }
-  }
+  def builder[Command, Event, State](statePredicate: FI.TypedPredicate[State]): CommandHandlerBuilder[Command, Event, State, State] =
```

Comment:
Should this not be akka.japi.function.Predicate or is java doc for TypedPredicate not accurate?

Reply:
Here I think `java.util.function.Predicate` can be used, or do we need the contravariant here?

```
public interface Predicate<T> {
  boolean test(T t);
```

vs

```
trait Predicate[-T] extends java.io.Serializable {
  def test(param: T): Boolean
}
```